### PR TITLE
Implement PickerAvoidingView and PickerStateProvider

### PR DESCRIPTION
**Pull request by jakub-trzebiatowski**
...which provide tools for ensuring that a picker is not covered by its own modal on iOS.

I'm still testing these changes, but they are ready to be reviewed, as I don't expect any major changes in the approach for now.

### CODE REVIEW

**Comment by jakub-trzebiatowski** [comment]
What's still to be done: creating a symmetric PR to `@react-native-picker/picker` and adding a relevant comment that these changes are subject to being reverted if this functionality is merged to the upstream.

File: src/constants.js
Comment (0xmiroslav) [inline src/constants.js:3]: can you also confirm this on iPhone SE, iPhone 14 Pro Max?

File: src/PickerAvoidingView/index.ios.js
Comment (0xmiroslav) [inline src/PickerAvoidingView/index.ios.js]: why do we need these values which are default already?

File: src/PickerAvoidingView/index.ios.js
Comment (0xmiroslav) [inline src/PickerAvoidingView/index.ios.js]: replace this with prop value (props.style)

File: src/PickerAvoidingView/index.ios.js
Comment (0xmiroslav) [inline src/PickerAvoidingView/index.ios.js]: `contentContainer` is better name.
and accept custom prop value (props.contentContainerStyle)

File: src/PickerAvoidingView/index.ios.js
Comment (0xmiroslav) [inline src/PickerAvoidingView/index.ios.js]: Similar to KeyboardAvoidingView, remove this here and let user define

File: src/PickerAvoidingView/index.js
Comment (0xmiroslav) [inline src/PickerAvoidingView/index.js]: return normal view similar to https://github.com/Expensify/App/blob/main/src/components/KeyboardAvoidingView/index.js

**Comment by 0xmiroslav** [comment]
@Julesssss please assign yourself as a reviewer

**Review by 0xmiroslav — CHANGES_REQUESTED** [review]

File: src/PickerAvoidingView/index.ios.js
Comment (jakub-trzebiatowski) [reply src/PickerAvoidingView/index.ios.js]: Habits from the web (as the defaults differ). Sorry.

File: src/PickerAvoidingView/index.ios.js
Comment (jakub-trzebiatowski) [reply src/PickerAvoidingView/index.ios.js]: Maybe let's add this when it's needed? As always, it can potentially break the solution, and we don't need it currently. If it will be requested in the upstream, it's another topic.

File: src/PickerAvoidingView/index.ios.js
Comment (jakub-trzebiatowski) [reply src/PickerAvoidingView/index.ios.js]: But in order for this component to work, i..e to do what's it's meant to do, it has to have some specific styling, right? Picker modal is on the bottom, so int needs to be a column, etc... Do you mean that some styles should be forced, but other should be accepted from the user?

File: src/PickerAvoidingView/index.js
Comment (jakub-trzebiatowski) [reply src/PickerAvoidingView/index.js]: Yup, if we support custom styles, then it's a must-have.

File: src/PickerAvoidingView/index.ios.js
Comment (0xmiroslav) [reply src/PickerAvoidingView/index.ios.js]: my idea was to exactly replicate KeyboardAvoidingView.
Let's say it's like normal View which doesn't have any custom prop set.
So I meant this:
```suggestion
        <View style={props.style}>
```

File: src/PickerAvoidingView/index.ios.js
Comment (0xmiroslav) [reply src/PickerAvoidingView/index.ios.js]: Actually why do we need to wrap with another View? Won't this just work?
```
<View style={props.style}
    {props.children}
    <Spacer />
</View>
```

File: src/PickerAvoidingView/index.ios.js
Comment (jakub-trzebiatowski) [reply src/PickerAvoidingView/index.ios.js]: But this isn't "a normal view", this is a component that solves the picker obstruction problem by providing a column layout and adding a child that provides the extra space as needed.

This behavior shouldn't be overridable, from my perspective. It shouldn't be possible to make this.. a row, or something else.

I can see some value in reflecting `KeyboardAvoidingView`, but what's even more important is to make something that works and can't be broken easily via the public interface it exposes.

Comment (0xmiroslav) [reply src/PickerAvoidingView/index.ios.js]: I like this solution: https://github.com/facebook/react-native/blob/4fdad9e765c45b5c60ce25e671f430e03edc9ef6/packages/react-native/Libraries/Components/Keyboard/KeyboardAvoidingView.js#L228-L237
```
          <View
            ref={this.viewRef}
            style={StyleSheet.compose(style, {paddingBottom: bottomHeight})}
            onLayout={this._onLayout}
            {...props}>
            {children}
          </View>
```

File: src/PickerAvoidingView/index.ios.js
Comment (jakub-trzebiatowski) [reply src/PickerAvoidingView/index.ios.js]: Just an intuition that it's safer to control the flexbox as a whole here, both the container and the items. In theory, the injected child/children could have some CSS style for flex items, which I didn't really want to support.

File: src/PickerAvoidingView/index.ios.js
Comment (jakub-trzebiatowski) [reply src/PickerAvoidingView/index.ios.js]: That's what I also figured out 😊

Comment (0xmiroslav) [reply src/PickerAvoidingView/index.ios.js]: ok let's try this, ignoring all previous comments and suggestions

File: src/constants.js
Comment (jakub-trzebiatowski) [reply src/constants.js:3]: Same! But a good idea to check.

Comment updated.

File: src/PickerAvoidingView/index.ios.js
Comment (jakub-trzebiatowski) [inline src/PickerAvoidingView/index.ios.js]: When I started implementing this, I thought that I actually needed it to be animated. But now, when I'm looking at this, it seems that it basically regressed to a `setTimeout`. I'll try simplifying it.

File: src/index.js
Comment (jakub-trzebiatowski) [inline src/index.js]: @0xmiroslav @Julesssss

To ensure that the "blank space" doesn't appear, I made scrolling aware of the duration of the modal animation.

This creates some cross-awareness between the scrolling to open picker (which is `react-native-picker-select`-specific) and `PickerAvoidingView`.

Maybe this means that a better target for this functionality in the upstream is actually `react-native-picker-select`?

File: src/PickerAvoidingView/index.ios.js
Comment (jakub-trzebiatowski) [reply src/PickerAvoidingView/index.ios.js]: Done

**Comment by 0xmiroslav** [comment]
index.ios.js:
```
    return (
        <View style={StyleSheet.compose(props.style, {paddingBottom: shouldAddSpace ? IOS_MODAL_HEIGHT : 0})}>
            {props.children}
        </View>
    );
```

index.js:
```
const PickerAvoidingView = (props) => {
    const viewProps = _.omit(props, ['enabled']);
    return (
        // eslint-disable-next-line react/jsx-props-no-spreading
        <View {...viewProps} />
    );
};
```

@cubuspl42 this is my suggestion

**Comment by 0xmiroslav** [comment]
If we animate like KeyboardAvoidingView, that would be a great bonus point.
So IOS_MODAL_HEIGHT value animates with same value and same duration as picker modal.
cc: @Julesssss 


https://user-images.githubusercontent.com/97473779/231738331-d2664e09-1521-43d4-aff2-14fb38c4ab17.mp4

**Comment by jakub-trzebiatowski** [comment]
> If we animate like KeyboardAvoidingView, that would be a great bonus point. So IOS_MODAL_HEIGHT value animates with same value and same duration as picker modal. cc: @Julesssss

Please note that the problem is of a bit different nature. Resizing the height of the bottom space does not scroll or move the content inside the `ScrollView`. What makes the picker element move over the modal is the `scrollTo` call.

**Comment by 0xmiroslav** [comment]
> > If we animate like KeyboardAvoidingView, that would be a great bonus point. So IOS_MODAL_HEIGHT value animates with same value and same duration as picker modal. cc: @Julesssss
> 
> Please note that the problem is of a bit different nature. Resizing the height of the bottom space does not scroll or move the content inside the `ScrollView`. What makes the picker element move over the modal is the `scrollTo` call.

I think you didn't get my point. It's not related to scroll.
My video is about resizing height, not scrolling to some view. Button is animating not because of scroll but resizing.
Scrolling should happen only after animation complete.

**Comment by 0xmiroslav** [comment]
@cubuspl42 I opened slack discussion to move things faster. Let's discuss there for further comments https://expensify.slack.com/archives/C01GTK53T8Q/p1681386144253199

File: src/index.js
Comment (jakub-trzebiatowski) [reply src/index.js]: @0xmiroslav @Julesssss 
There a few benefits:
* The `PickerAvoidingView` and the scrolling logic are somewhat aware of each other
* `react-native-picker-select` is our direct picker dependency. It used not to depend on `@react-native-picker/picker` and _theoretically_ it might not depend on it again (although unlikely)
* We already have a fork, so I can just select these commits and submit them as a PR

Unless you oppose, I'll proceed with creating an upstream PR in `react-native-picker-select` 🙃

File: src/PickerAvoidingView/index.ios.js
Comment (Julesssss) [reply src/PickerAvoidingView/index.ios.js]: Thanks for the additional context. Maybe we could add a short comment explaining the reasoning behind this to make it obvious to others?

File: src/index.js
Comment (Julesssss) [reply src/index.js]: Thanks, yeah I think your suggestion makes sense to me. Better to separate the two libs to keep the implementation simpler.

File: src/PickerAvoidingView/index.ios.js
Comment (jakub-trzebiatowski) [reply src/PickerAvoidingView/index.ios.js]: I followed the advice by @0xmiroslav to use the `paddingBottom` instead. So I'm resolving this thread.

File: src/index.js
Comment (jakub-trzebiatowski) [reply src/index.js]: ~[PR filed](https://github.com/Expensify/react-native-picker-select/pull/11)!~

It's stacked on top of one of our PRs, from January. Who knows, maybe the maintiner will wake up 😒

Comment (Julesssss) [reply src/index.js]: We can still review/merge our fork for now. 

Also, I'm trying to offer the maintainer money in order to review our PR. But I'm struggling to find any contact details.

Comment (jakub-trzebiatowski) [reply src/index.js]: @Julesssss That PR was meant for the upstream, but I just figured out that  filed it to our fork 🙈 That's how doing things late in the evening ends.

Comment (jakub-trzebiatowski) [reply src/index.js]: [Upstream PR](https://github.com/lawnstarter/react-native-picker-select/pull/507)

File: src/index.js
Comment (jakub-trzebiatowski) [inline src/index.js]: @Julesssss

> Can we figure out a way of avoiding the setTimeout? Possible via a hook?

Comment (jakub-trzebiatowski) [reply src/index.js]: Are you concerned about the component being unmounted before the timeout finishes?

Comment (Julesssss) [reply src/index.js]: We just try to avoid using `setTimeout` as a pattern if possible

https://github.com/Expensify/App/blob/ab9b7d9ef1fb1656f5ccfc28ca4a68820fa373b4/contributingGuides/PR_REVIEW_GUIDELINES.md?plain=1#L56

Comment (jakub-trzebiatowski) [reply src/index.js]: Ok, so [this commit](https://github.com/Expensify/react-native-picker-select/pull/8/commits/eeb3bff23ef3d060148057263466aa674a5fadcc) wraps that timeout in a hook, preventing it from occurring when the component is unmounted.

The thing is that we can't scroll until the layout hasn't been updated, and the layout isn't updated until the animation finishes, not to cause a "blink" (as [requested here](https://github.com/Expensify/App/pull/17279#discussion_r1164239617) by @0xmiroslav).

>  Make sure any `setTimeout`s are absolutely necessary, not just a workaround for a situation that isn’t fully understood (e.g. a race condition).

The second part is definitely not the case here. The situation is fully understood. The first part is arguable. It's not **absolutely** necessary to use a time-based solution here, but creating a communication channel between `PickerAvoidingView` and the picker itself would be quite a big effort here.

Both `PickerAvoidingView` and `scrollToInput` share the `IOS_MODAL_ANIMATION_DURATION_MS` constant, so their cross-dependence is clearly marked in the code.

File: src/index.js
Comment (jakub-trzebiatowski) [reply src/index.js]: @Julesssss 
> Also, I'm trying to offer the maintainer money in order to review our PR. But I'm struggling to find any contact details.

The person probably currently administrating the [LawnStarter organization](https://github.com/lawnstarter):
https://github.com/tomschlick
https://twitter.com/tomschlick
tom (at) schlick.email

Author/maintainer who used to work for LawnStarter till 2021:
https://github.com/lfkwtz
https://www.linkedin.com/in/lefkowitzmichael/
lefkowitz.michael (at) gmail.com

Comment (Julesssss) [reply src/index.js]: Ah, thanks. I had reached out on Linkedin but couldn't find their emails anywhere! I just sent a message.

File: src/index.js
Comment (jakub-trzebiatowski) [reply src/index.js]: @Julesssss What's your perspective on this? I tried to explain why I believe that a timer-based solution does not indicate a non-fully understood situation here. While the hook-based solution doesn't remove the timer completely, it improves how the timer is managed, by coupling it to the component lifecycle (canceling on umount). Does this look acceptable?

I believe that this is the only real topic left in this pair of PRs. It really would be great to start closing this, I think!

Comment (Julesssss) [reply src/index.js]: Honestly I'm not 100% sure what's best here, but you make a good case. What do you think @0xmiroslav ?

**Comment by 0xmiroslav** [comment]
@cubuspl42 please fix conflict and point correct hash on app PR

File: src/index.js
Comment (jakub-trzebiatowski) [reply src/index.js]: @0xmiroslav writes [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1681852803227919?thread_ts=1681386144.253199&cid=C01GTK53T8Q):

> If timeout removal is that complex and no other solution, then fine

I would say that there is no simple way to drop the timeout-based approach here. I'll resolve the thread.

File: package.json
Comment (0xmiroslav) [inline package.json]: Why is this change needed?
If it's not intentional, revert back.

File: src/index.js
Comment (0xmiroslav) [inline src/index.js]: Is there any other way to avoid this wrapper?

File: package.json
Comment (jakub-trzebiatowski) [reply package.json]: Hooks were added to React in version 16.8.

Comment (0xmiroslav) [reply package.json]: ok, but can we find other way without using hooks?

File: src/index.js
Comment (jakub-trzebiatowski) [reply src/index.js]: Hooks can't be used in class components. I used `useEffect` hook, indirectly via a helper functional component, to ensure that no effect is run after our component gets unmounted. Do you think this should be done manually in the component lifecycle methods?

Comment (0xmiroslav) [reply src/index.js]: > Do you think this should be done manually in the component lifecycle methods?

yes, we should try to avoid complex code as possible

Comment (jakub-trzebiatowski) [reply src/index.js]: `useEffect` logic, implemented using component lifecycle methods, looks more or less like this:

```js
componentDidMount() {
    this.performEffect();
  }

  componentDidUpdate(prevProps) {
    // check if dependencies changed
    if (/* dependencies changed */) {
      // Clean up the previous effect before applying the new one
      if (this.cleanupEffect) {
        this.cleanupEffect();
      }
      this.performEffect();
    }
  }

  componentWillUnmount() {
    if (this.cleanupEffect) {
      this.cleanupEffect();
    }
  }

  performEffect() {
    // effect code here

    // store the cleanup function as an instance variable
    this.cleanupEffect = () => {
      // cleanup/cancelling code here
    };
  }
```

Is that the simple version?

Comment (jakub-trzebiatowski) [reply src/index.js]: Hm, maybe I'm missing some simpler options. I do agree that things should be as simple as possible. Would you share what is on your mind? You can see the problem we're solving. How would you approach that?

Comment (0xmiroslav) [reply src/index.js]: Isn't that enough to clearTimout on componentWillUnmount?

Comment (jakub-trzebiatowski) [reply src/index.js]: @0xmiroslav Let me know if you consider [something like this](https://github.com/cubuspl42/react-native-picker-select/commit/0d097d7856a088ebea3e208d32f50a3536dd7276) simpler.

Comment (0xmiroslav) [reply src/index.js]: We can directly use setTimeout. no need `schedule`. It's also using setTimeout
What @Julesssss meant `hook` was to use something like `onModalShow ` callback.

Comment (0xmiroslav) [reply src/index.js]: did you already try `onModalShow` callback?

Comment (jakub-trzebiatowski) [reply src/index.js]: By "there's no need for `schedule`", you're saying that you wouldn't cancel this timeout in any circumstances, right? So you would guess that it wouldn't cause any trouble if it was run after the picker component is unmounted, right?

I'm not against it, I'm just trying to ensure that I read your words as you mean them. I haven't even tested such edge case, as it also wouldn't be very easy to test. It's just a good pattern to cancel all cancellable side effects when the component which "owns" them is unmounted, destroyed, or something like that.

Comment (0xmiroslav) [reply src/index.js]: > you wouldn't cancel this timeout in any circumstances

I asked: "Isn't that enough to clearTimout on componentWillUnmount?"

Also did you try `onModalShow` callback?

Comment (cubuspl42) [reply src/index.js:669]: > did you already try `onModalShow` callback?

No, I wasn't really aware of it. I'll check it.

Comment (jakub-trzebiatowski) [reply src/index.js]: > I asked: "Isn't that enough to clearTimout on componentWillUnmount?"

Sorry, I missed that part. It's not really the same, as `useEffect` also cancels the effect when the dependencies change. But maybe that part could be dropped.

I'm just investigating the possibility of using [`onShow`](https://reactnative.dev/docs/modal) function. `onModalShow` is a method from [a library](https://github.com/react-native-modal/react-native-modal) that `react-native-picker-select` doesn't use. It looks promising and I believe that it's a very good suggestion.

Comment (jakub-trzebiatowski) [reply src/index.js]: @0xmiroslav I replaced the time-based approach with `onShow`, like you suggested! I changed the name/contract of the flag stored in `PickerStateContext`, but that's semantics. I think that the whole solution actually simplified.

File: package.json
Comment (jakub-trzebiatowski) [reply package.json]: Yes, we can replace all hooks with non-hook concepts. Are you sure that it's unacceptable to start using hooks in this repository?

Comment (0xmiroslav) [reply package.json]: Upgrading react version is out of scope. Should not be handled here as long as we're not using any hooks.

Comment (jakub-trzebiatowski) [reply package.json]: Nevermind, it's just one use currently. I'll just drop it, if you think it's a good idea.

Comment (jakub-trzebiatowski) [reply package.json]: Replaced with a class component

File: src/index.js
Comment (0xmiroslav) [inline src/index.js:250]: can you share 2 videos to clarify difference?

File: src/index.js
Comment (0xmiroslav) [inline src/index.js]: It makes more sense to move this logic to `triggerOpenCloseCallbacks`

File: src/index.js
Comment (jakub-trzebiatowski) [reply src/index.js:250]: It's just a small padding, which is "accidentally" the same as we use in Expensify. It just looks a bit better.

After the scroll happens, this is how it looks now:

image

How it looked before:

image

Comment (jakub-trzebiatowski) [reply src/index.js:250]: Relevant line in Expensify: https://github.com/Expensify/App/blob/main/src/components/Form.js#L357

File: src/index.js
Comment (jakub-trzebiatowski) [reply src/index.js]: With changing the name of `triggerOpenCloseCallbacks`? Because now it says "trigger open & close callback"

Comment (jakub-trzebiatowski) [reply src/index.js]: I'll change it to `triggerCallbacks`

Comment (0xmiroslav) [reply src/index.js]: That naming is fine I think

Comment (0xmiroslav) [reply src/index.js]: do you have a better naming?

Comment (jakub-trzebiatowski) [reply src/index.js]: Moved and renamed

Comment (jakub-trzebiatowski) [reply src/index.js]: Well, it's just a simple observation that that name is very specific, and we start doing something else than _just_ calling open & close callbacks. So I just renamed to `triggerCallback` (as, let's say, that `setIsModalOpen` is some _kind_ of callback, stretching the definition a bit).

It's no big deal.

**Review by 0xmiroslav — DISMISSED** [review]
Looks good! Checklist will be done in app PR.
For now, please point latest commit hash in app PR.
@Julesssss will be back from OOO tomorrow and after this PR merged, we can simply update commit hash to merged one.

File: src/PickerAvoidingView/index.ios.js
Comment (jakub-trzebiatowski) [inline src/PickerAvoidingView/index.ios.js]: @0xmiroslav I found a small oversight. In the previous revision, I haven't forwarded other props to the `View`, as we do on the non-iOS implementation. This is now fixed.

File: src/PickerAvoidingView/index.ios.js
Comment (jakub-trzebiatowski) [inline src/PickerAvoidingView/index.ios.js:32]: I changed this to the `Consumer` API, as `useContext` is also React 16.8+. As we decided not to bump the React version, this is a necessary change. I oversaw this earlier.

File: src/PickerAvoidingView/index.js
Comment (0xmiroslav) [inline src/PickerAvoidingView/index.js]:
```suggestion
    return <View {...viewProps} />;
```

File: src/PickerAvoidingView/index.ios.js
Comment (0xmiroslav) [inline src/PickerAvoidingView/index.ios.js]:
```suggestion
        const { enabled, style, ...viewProps } = this.props;
```

File: src/PickerAvoidingView/index.ios.js
Comment (0xmiroslav) [inline src/PickerAvoidingView/index.ios.js]:
```suggestion
                    return <View style={effectiveStyle} {...viewProps}>
```

**Review by 0xmiroslav — CHANGES_REQUESTED** [review]

File: src/PickerAvoidingView/index.ios.js
Comment (0xmiroslav) [inline src/PickerAvoidingView/index.ios.js:25]: Please add propType for `enabled`

**Review by 0xmiroslav — APPROVED** [review]

File: src/constants.js
Comment (Julesssss) [reply src/constants.js:3]: What about iPad?

**Review by Julesssss — COMMENTED** [review]
Looking good, just curious if the hardcoded height is suitable for iPad devices.

**Comment by 0xmiroslav** [comment]
> Looking good, just curious if the hardcoded height is suitable for iPad devices.

@Julesssss Portrait mode exactly matches hardcoded height but landscape mode height is smaller.
That hardcoded height (262) was introduced in https://github.com/Expensify/react-native-picker-select/pull/3
There was already discussion about that but didn't care about iPad landscape mode - https://github.com/Expensify/react-native-picker-select/pull/3/files#r1046091730 cc: @tgolen @chrispader

So 3 options here to go ahead
- find solution to get modal height programatically
- if it's iPad landscape, return xxx (measured value), if iPad portrait or iPhone, return 262
- leave as is now given that iPad is not officially supported platform

I tested on iPad mini 6th generation
Simulator Screen Shot - iPad mini (6th generation) - 2023-04-25 at 11 48 31
Simulator Screen Shot - iPad mini (6th generation) - 2023-04-25 at 11 48 47

NOTE: Currently team doesn't really care about iPad responsive.
i.e. Login page style broken but such [issue](https://github.com/Expensify/App/issues/16612) was closed given it's not officially supported platform.

https://user-images.githubusercontent.com/97473779/234248711-c4be4b87-f6a6-4af6-a04a-1662967a49da.mov

**Comment by jakub-trzebiatowski** [comment]
@Julesssss @0xmiroslav It's a good catch! Still, I think that this PR doesn't really introduce any regression in the discussed matter. The constant used for the modal was already present in the source code. Taking into consideration that @0xmiroslav is saying that even the sign in page doesn't support iPad properly, and iPad doesn't get first-class support, I think we don't have to improve the modal measurements, as the effect on iPad is not _that_ bad.

**Comment by Julesssss** [comment]
Yeah this looks fine, so I agree we shouldn't care about this for now 👍

**Review by Julesssss — APPROVED** [review]

File: src/index.js
Comment (s77rt) [inline src/index.js:499]: If we show the Modal, `onShow` callback will be called and set modal as shown (`isModalShown=true`). Now if we unmount that modal (while it's shown) the state of the modal being shown won't be reset accordingly - state is stuck at `true`. We should have cleared the state on `componentWillUnmount`. (Coming from https://github.com/Expensify/App/issues/23044)

**Pull request merged**
